dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java

```diff
@@ -60,16 +64,26 @@ public List<BlockWorkerInfo> getPreferredWorkers(WorkerClusterView workerCluster
           "Not enough workers in the cluster %d workers in the cluster but %d required",
           workerClusterView.size(), count));
     }
-    List<BlockWorkerInfo> blockWorkerInfoList = workerClusterView.stream()
-        .map(w -> new BlockWorkerInfo(w.getIdentity(), w.getAddress(), w.getCapacityBytes(),
-            w.getUsedBytes(), w.getState() == WorkerState.LIVE))
-        .collect(Collectors.toList());
-    HASH_PROVIDER.refresh(blockWorkerInfoList, mNumVirtualNodes);
-    List<BlockWorkerInfo> workers = HASH_PROVIDER.getMultiple(fileId, count);
+    Set<WorkerIdentity> workerIdentities = workerClusterView.workerIds();
+    mHashProvider.refresh(workerIdentities, mNumVirtualNodes);
+    List<WorkerIdentity> workers = mHashProvider.getMultiple(fileId, count);
     if (workers.size() != count) {
       throw new ResourceExhaustedException(String.format(
           "Found %d workers from the hash ring but %d required", workers.size(), count));
     }
-    return workers;
+    ImmutableList.Builder<BlockWorkerInfo> builder = ImmutableList.builder();
+    for (WorkerIdentity worker : workers) {
+      WorkerInfo workerInfo = workerClusterView.getWorkerById(worker)
+          .orElseThrow(() -> new IllegalStateException(
+              String.format("Hash provider returned a non-existent worker: %s, "
```

Comment:
i feel not comfortable just throwing an exception here. this might cause the upstream application to fail

Reply:
updated